File: Data Structures & Algorithms/trapping-rain-water/submission-9.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        lm, rm = height[l], height[r]
        res = 0
        while l < r:
            if height[l] < height[r]:
                l += 1
                lm = max(height[l], lm)
                res += lm - height[l]
            else:
                r -= 1
                rm = max(height[r], rm)
                res += rm - height[r]
        return res

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-9.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> freqMap;
        int ret = 0;
        int l = 0;
        int largest = 0;
        for (int r = 0; r < s.size(); ++r) {
            ++freqMap[s.at(r)];
            largest = max(largest, freqMap[s[r]]);
            while ((r - l + 1) - largest > k) {
                --freqMap[s.at(l)];
                ++l;
            }
            ret = max(ret, r - l + 1);
        }
        return ret;
    }
};
